src/app/Plans/components/helpers.ts

```diff
@@ -1,33 +1,133 @@
   );
   return condition ? PlanStatusDisplayType[condition.type] : '';
 };
 
-// TODO maybe generalize this for cold migrations too
-type WarmPlanState =
-  | 'NotStarted'
-  | 'Starting'
-  | 'Copying'
-  | 'AbortedCopying'
-  | 'StartingCutover'
-  | 'Cutover'
-  | 'Finished';
+export const getMigStatusState = (plan: IPlan) => {
+  const conditions = plan.status?.conditions || [];
+  let title = '';
+  let variant: ProgressVariant | undefined;
+
+  switch (true) {
+    case hasCondition(conditions, 'Executing') && plan.spec.warm: {
+      title = 'Running cutover';
+      break;
+    }
+    case hasCondition(conditions, 'Executing'): {
+      title = 'Running';
+      break;
+    }
+    case hasCondition(conditions, 'Succeeded'): {
+      title = 'Succeeded';
+      variant = ProgressVariant.success;
+      break;
+    }
+    case hasCondition(conditions, 'Canceled'): {
+      title = 'Canceled';
+      break;
+    }
+    case hasCondition(conditions, 'Failed'): {
+      title = 'Failed';
+      variant = ProgressVariant.danger;
+      break;
+    }
+    case !!plan.status?.migration?.started: {
+      if (canPlanBeStarted(plan)) {
+        title = 'Finished - Incomplete';
+        variant = ProgressVariant.warning;
+      } else {
+        console.warn('Migration plan unexpected status:', plan);
+      }
+      break;
+    }
+    default: {
+      title = '';
+    }
+  }
+
+  return {
+    title,
+    variant,
+  };
+};
 
-export const getWarmPlanState = (
+export const getButtonState = (
+  state: PlanState | null,
+  plan: IPlan,
+  migration: IMigration | null
+): PlanActionButtonType | null => {
+  let type: PlanActionButtonType | null = null;
+
+  switch (true) {
+    case state === 'NotStarted': {
+      type = 'Start';
+      break;
+    }
+    case plan.status?.migration?.started && canPlanBeStarted(plan): {
+      type = 'Restart';
+      break;
+    }
+    case state === 'Starting': {
+      type = null; // spinner
+      break;
+    }
+    case state === 'Starting' && !migration?.spec.cutover: {
+      type = 'Cutover'; // spinner
+      break;
+    }
+    case state === 'Copying': {
+      type = 'Cutover';
+      break;
+    }
+    case state === 'AbortedCopying': {
+      type = 'Restart';
+      break;
+    }
+    case state === 'StartingCutover': {
+      type = null; // spinner
+      break;
+    }
+    // "Cutover" name is warm-specific, but this state applies to both warm and cold
+    // it could just be called 'Migrating' or 'PipelineRunning'
+    case state === 'Cutover': {
+      type = null; // no button
+      break;
+    }
+    case state === 'Finished': {
+      type = null; // no button
+      break;
+    }
+    case plan.status?.conditions &&
+      hasCondition(plan.status?.conditions, 'Ready') &&
+      !plan.status?.migration?.started: {
+      type = 'Start';
+      break;
+    }
+    default: {
+      type = null;
+    }
+  }
+
+  return type;
+};
+
+export const getPlanState = (
```

Comment:
On line 135 here in `getPlanState`, you return null if `!plan.spec.warm`. This means `planState` is null for every single cold plan. Interestingly enough this doesn't seem to be affecting any of the mocked plans in the preview, because:

1. the only two cases in the Status column we show are the `<StatusIcon>` case (which you also show if `!plan.status?.migration?.started`) and the `<Progress>` case (which is the base case, it's showing for all the other plans because `planState` is null)
2. the Progress case uses `getMigStatusState` which isn't using the `planState` at all anyway.
3. the `getButtonState` function is using `planState`, but the only 2 cases it's hitting for cold plans are `'Restart'` (which you're also using `plan.status?.migration?.started` for), `Start` (which you're also using `plan.status.conditions` for in a separate case from the NotStarted one) and null, which is your default case anyway.

But it will behave incorrectly in other cases we aren't mocking: for example, we'll no longer get a "Preparing for migration" case when a cold plan would be in the `Starting` state. We may want to add a mock plan to show this state.

You should remove the `!plan.spec.warm` condition from `getPlanState` so that you actually have a defined `planState` for every plan, and then I would think we should drive the `getButtonState` and `getMigStatusState` entirely from `planState` and not touch any of the `plan.status` stuff in it if we can help it, so all of that logic is in `getPlanState`.

We should also probably remove all the usages of `isPlanBeingStarted` and drive those all from the `'Starting'` planState. Single source of truth, single source of plan state logic :)

Reply:
It does actually look like some of the logic won't work right for cold plans as well. Any cold plan with the Executing condition will end up in the Starting state.. you may want to have that whole `if` block affect only `plan.spec.warm` plans, and have another one that puts cold plans in the `RunningPipeline` state (or whatever we rename `Cutover` to).

Reply:
Ok I _think_ I addressed most all of this. There's still a couple of places in the code that I think could be cleaned up a little further, take a look when you get time and let me know what you think!

Reply:
Resolving this because you got most of it, will open another thread for new comments.